osi3trace/osi_trace.py

```diff
@@ -43,9 +49,156 @@ def map_message_type(type_name):
+
+    def from_file(self, path, type_name="SensorView", cache_messages=False, topic=None):
+        """
+        Initializes the trace reader depending on the trace file format.
+        
+        Args:
+            path (str): The path to the trace file.
+            type_name (str): The type name of the messages in the trace; check supported message types with `OSITrace.message_types()`.
+            cache_messages (bool): Whether to cache messages in memory (only applies to single-channel traces).
+            topic (str): The topic name for multi-channel traces (only applies to multi-channel traces); Using the first available topic if not specified.
+        """
+        self.reader = self._init_reader(Path(path), type_name, cache_messages, topic)
+
+    def restart(self, index=None):
+        """
+        Restart the trace reader.
+
+        Note:
+            Multi-channel traces don't support restarting from a specific index.
+        """
+        return self.reader.restart(index)
+
+    def __iter__(self):
+        return self.reader.__iter__()
+
+    def close(self):
+        return self.reader.close()
 
+    def retrieve_offsets(self, limit=None):
+        if isinstance(self.reader, OSITraceSingle):
+            return self.reader.retrieve_offsets(limit)
+        raise NotImplementedError("Offsets are only supported for single-channel traces.")
+
+    def retrieve_message(self, index=None, skip=False):
+        if isinstance(self.reader, OSITraceSingle):
+            return self.reader.retrieve_message(index, skip)
+        raise NotImplementedError("Index-based message retrieval is only supported for single-channel traces.")
+
+    def get_message_by_index(self, index):
+        if isinstance(self.reader, OSITraceSingle):
+            return self.reader.get_message_by_index(index)
+        raise NotImplementedError("Index-based message retrieval is only supported for single-channel traces.")
+
+    def get_messages_in_index_range(self, begin, end):
+        if isinstance(self.reader, OSITraceSingle):
+            return self.reader.get_messages_in_index_range(begin, end)
+        raise NotImplementedError("Index-based message retrieval is only supported for single-channel traces.")
+
+    def get_available_topics(self):
+        if isinstance(self.reader, OSITraceMulti):
+            return self.reader.get_available_topics()
+        raise NotImplementedError("Getting available topics is only supported for multi-channel traces.")
+    
+    def get_file_metadata(self):
+        if isinstance(self.reader, OSITraceMulti):
+            return self.reader.get_file_metadata()
+        raise NotImplementedError("Getting file metadata is only supported for multi-channel traces.")
+    
+    def get_channel_metadata(self):
+        if isinstance(self.reader, OSITraceMulti):
+            return self.reader.get_channel_metadata()
+        raise NotImplementedError("Getting channel metadata is only supported for multi-channel traces.")
```

Comment:
Probably these should be pushed down into the implementation classes, where they can more naturally error out (by not implementing the methods), and can be implemented in the future going forward. Or in other words, if we are using inheritance from ReaderBase anyways, we might as well let method dispatch handle our non-implementation choices.